ddtrace/sampler.py

```diff
@@ -92,3 +103,266 @@ def set_sample_rate_by_service(self, rate_by_service):
+        :param resource: Rule to match the `span.resource` on, default no rule defined
+        :type resource: :obj:`object` to directly compare, :obj:`function` to evaluate, or :class:`re.Pattern` to match
+        :param tags: Dictionary of tag -> pattern to match on the `span.meta`, default no rule defined
+        :type tags: :obj:`dict` mapping :obj:`str` tag names to either :obj:`object` to directly compare,
+            :obj:`function` to evaluate, or :class:`re.Pattern` to match
+        """
+        # Enforce sample rate constraints
+        if not 0.0 <= sample_rate <= 1.0:
+            raise ValueError(
+                'SamplingRule(sample_rate={!r}) must be greater than or equal to 0.0 and less than or equal to 1.0',
+            )
+
+        self.sample_rate = sample_rate
+        self.service = service
+        self.name = name
+        self.resource = resource
+
+        if tags is not self.NO_RULE and not isinstance(tags, dict):
+            raise TypeError('SamplingRule(tags={!r}) must be a dict')
+        self.tags = tags
+
+    @property
+    def sample_rate(self):
+        return self._sample_rate
+
+    @sample_rate.setter
+    def sample_rate(self, sample_rate):
+        self._sample_rate = sample_rate
+        self._sampling_id_threshold = sample_rate * MAX_TRACE_ID
+
+    def _pattern_matches(self, prop, pattern):
+        # If the rule is not set, then assume it matches
+        # DEV: Having no rule and being `None` are different things
+        #   e.g. ignoring `span.service` vs `span.service == None`
+        if pattern is self.NO_RULE:
+            return True
+
+        # If the pattern is callable (e.g. a function) then call it passing the prop
+        #   The expected return value is a boolean so cast the response in case it isn't
+        if callable(pattern):
+            try:
+                return bool(pattern(prop))
+            except Exception as e:
+                log.warning('%r pattern %r failed with %r: %s', self, pattern, prop, e)
+                # Their function failed to validate, assume it is a False
+                return False
+
+        # The pattern is a regular expression and the prop is a string
+        if isinstance(pattern, pattern_type):
+            try:
+                return bool(pattern.match(str(prop)))
+            except (ValueError, TypeError) as e:
+                # This is to guard us against the casting to a string (shouldn't happen, but still)
+                log.warning('%r pattern %r failed with %r: %s', self, pattern, prop, e)
+                return False
+
+        # Exact match on the values
+        return prop == pattern
+
+    def _tags_match(self, span):
+        # No rule was set, then assume it matches
+        if self.tags is self.NO_RULE:
+            return True
+
+        # No tags were provided to check, so it passes
+        if not self.tags:
+            return True
```

Comment:
That shouldn't be needed, right?

```python
>>> all([])
True
```

Reply:
This was more to guard against any falsey value, e.g. if `self.tags = None` somehow.

Reply:
Alright, so that links back to my `NO_RULE` comment. `NO_RULE` is definitely not useful for `tags` since it's expected to be a list.

I'm not a fan of this kind of check I guess in this case that's good enough.